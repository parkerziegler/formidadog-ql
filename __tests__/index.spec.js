"use strict";

const { graphql } = require("graphql");
const fetch = require("isomorphic-fetch");
require("dotenv").config();

const { schema } = require("../src/schema");

describe("formidadog-ql server", () => {
  /**
   * Fetch dogs from dogs.formidable.dev since our schema initalization
   * requires that they be available at runtime. The JSON file is small
   * enough that it shouldn't impact test time too drastically.
   */
  let dogs;
  beforeAll(async () => {
    const res = await fetch("https://dogs.formidable.dev/dogs");
    const data = await res.json();
    dogs = data.map((dog) => ({
      ...dog,
      likes: 0,
      pats: 0,
      treats: 0,
      bellyscratches: 0,
    }));
  });

  it("executes the dogs query", async () => {
    const query = `
      query dogs {
        dogs {
          key
          name
          breed
          color
          imageUrl
          description
          likes
        }
      }
    `;

    const result = await graphql(schema(dogs), query);
    expect(result).toMatchSnapshot();
  });

  it("executes the dog query", async () => {
    const query = `
      query dog($key: ID!) {
        dog(key: $key) {
          breed
          name
        }
      }
    `;

    const result = await graphql(schema(dogs), query, undefined, undefined, {
      key: "VmeRTX7j-",
    });

    expect(result).toMatchSnapshot();
    expect(result.data.dog.name).toEqual("Dixie");
  });

  it("executes the likeDog mutation", async () => {
    const mutation = `
      mutation likeDog($key: ID!) {
        likeDog(key: $key) {
          likes
        }
      }
    `;

    const result = await graphql(schema(dogs), mutation, undefined, undefined, {
      key: "VmeRTX7j-",
    });

    expect(result).toMatchSnapshot();
    expect(result.data.likeDog.likes).toEqual(1);
  });

  it("executes the likeAllDogs mutation", async () => {
    const mutation = `
      mutation likeAllDogs {
        likeAllDogs {
          likes
        }
      }
    `;

    const result = await graphql(schema(dogs), mutation);

    expect(result).toMatchSnapshot();
    expect(result.data.likeAllDogs.every((dog) => dog.likes > 0)).toBe(true);
  });

  it("executes the patDog mutation", async () => {
    const mutation = `
      mutation patDog($key: ID!) {
        patDog(key: $key) {
          pats
        }
      }
    `;

    const result = await graphql(schema(dogs), mutation, undefined, undefined, {
      key: "VmeRTX7j-",
    });

    expect(result).toMatchSnapshot();
    expect(result.data.patDog.pats).toEqual(1);
  });

  it("executes the patAllDogs mutation", async () => {
    const mutation = `
      mutation patAllDogs {
        patAllDogs {
          pats
        }
      }
    `;

    const result = await graphql(schema(dogs), mutation);

    expect(result).toMatchSnapshot();
    expect(result.data.patAllDogs.every((dog) => dog.pats > 0)).toBe(true);
  });

  it("executes the treatDog mutation", async () => {
    const mutation = `
      mutation treatDog($key: ID!) {
        treatDog(key: $key) {
          treats
        }
      }
    `;

    const result = await graphql(schema(dogs), mutation, undefined, undefined, {
      key: "VmeRTX7j-",
    });

    expect(result).toMatchSnapshot();
    expect(result.data.treatDog.treats).toEqual(1);
  });

  it("executes the treatAllDogs mutation", async () => {
    const mutation = `
      mutation treatAllDogs {
        treatAllDogs {
          treats
        }
      }
    `;

    const result = await graphql(schema(dogs), mutation);

    expect(result).toMatchSnapshot();
    expect(result.data.treatAllDogs.every((dog) => dog.treats > 0)).toBe(true);
  });

  it("executes the bellyscratchDog mutation", async () => {
    const mutation = `
      mutation bellyscratchDog($key: ID!) {
        bellyscratchDog(key: $key) {
          bellyscratches
        }
      }
    `;

    const result = await graphql(schema(dogs), mutation, undefined, undefined, {
      key: "VmeRTX7j-",
    });

    expect(result).toMatchSnapshot();
    expect(result.data.bellyscratchDog.bellyscratches).toEqual(1);
  });

  it("executes the bellyscratchAllDogs mutation", async () => {
    const mutation = `
      mutation bellyscratchAllDogs {
        bellyscratchAllDogs {
          bellyscratches
        }
      }
    `;

    const result = await graphql(schema(dogs), mutation);

    expect(result).toMatchSnapshot();
    expect(
      result.data.bellyscratchAllDogs.every((dog) => dog.bellyscratches > 0)
    ).toBe(true);
  });
});
