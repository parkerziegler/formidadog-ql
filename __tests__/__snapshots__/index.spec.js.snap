// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formidadog-ql server executes the bellyscratchAllDogs mutation 1`] = `
Object {
  "data": Object {
    "bellyscratchAllDogs": Array [
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 2,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
      Object {
        "bellyscratches": 1,
      },
    ],
  },
}
`;

exports[`formidadog-ql server executes the bellyscratchDog mutation 1`] = `
Object {
  "data": Object {
    "bellyscratchDog": Object {
      "bellyscratches": 1,
    },
  },
}
`;

exports[`formidadog-ql server executes the dog query 1`] = `
Object {
  "data": Object {
    "dog": Object {
      "breed": "Pit Mix",
      "name": "Dixie",
    },
  },
}
`;

exports[`formidadog-ql server executes the dogs query 1`] = `
Object {
  "data": Object {
    "dogs": Array [
      Object {
        "breed": "Chihuahua",
        "color": "Tan",
        "description": "This sassy little dog has a super-size personality. He knows what he wants and goes after it with single-minded determination. For his size, he’s an excellent watchdog, but he can be yappy if he’s not taught to moderate his barking.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/milo.jpg",
        "key": "r17cRT2uW",
        "likes": 0,
        "name": "Milo",
      },
      Object {
        "breed": "Labrador Retriever",
        "color": "Chocolate",
        "description": "The Labrador Retriever was bred to be both a friendly companion and a useful working dog breed. Historically, he earned his keep as a fisherman’s helper: hauling nets, fetching ropes, and retrieving fish from the chilly North Atlantic.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/beau.jpg",
        "key": "H1jyJC2_b",
        "likes": 0,
        "name": "Beau",
      },
      Object {
        "breed": "Pit Mix",
        "color": "Brown",
        "description": "Pit Bull dogs have historically been known as courageous and heroic animals. Pit Bulls are naturally friendly, active dogs that need a lot of exercise. Most Pits are patient and adore children, which makes them wonderful family dogs.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/dixie.jpg",
        "key": "VmeRTX7j-",
        "likes": 0,
        "name": "Dixie",
      },
      Object {
        "breed": "Goldendoodle",
        "color": "Tan",
        "description": "Goldendoodle is a hybrid of a Golden Retriever and a Standard Poodle. While Labradoodles were bred to be guide dogs for vision-impaired people with dog-sensitive allergies, Goldendoodles have been bred primarily to be family pets, although they can also make terrific service dogs.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/leo.jpg",
        "key": "rJKI1Rhu-",
        "likes": 0,
        "name": "Leo",
      },
      Object {
        "breed": "Mutt with Border Collie DNA",
        "color": "Black",
        "description": "Sophie does tricks, enthusiastic barker, fan of the boss’s organic salmon, diagnosed OCD.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/sophie.jpg",
        "key": "tEQM4Lzj_",
        "likes": 0,
        "name": "Sophie",
      },
      Object {
        "breed": "Chihuahua",
        "color": "Brown",
        "description": "Fourteen pounds of speedy, food-seeking, barkiness.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/rusty.jpg",
        "key": "cLnG8C2d_",
        "likes": 0,
        "name": "Rusty",
      },
      Object {
        "breed": "Beagle",
        "color": "White and Brown",
        "description": "Beagles are scenthounds, meaning they live to use their nose. They’re a comfortable size to tote around in your car, simple to groom, and their exercise needs are easily met with a long, meandering walk that gives them plenty of time to sniff.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/lucy.jpg",
        "key": "Sk1GkChuZ",
        "likes": 0,
        "name": "Lucy",
      },
      Object {
        "breed": "West Highland White Terrier",
        "color": "White",
        "description": "You may recognize the Westie from her long-running stint as the mascot for Cesar pet food, but she’s more than just a cute face. A true terrier, she’s a fast and clever hunter, plus her lighthearted nature makes for a pet who’s always game for some fun.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/callie.jpg",
        "key": "HJMVJR3ub",
        "likes": 0,
        "name": "Callie",
      },
      Object {
        "breed": "Pomeranian",
        "color": "Brown",
        "description": "Tiny and bright-eyed, the spunky Pomeranian greets the world with endless curiosity and a sure sense that he’s the cutest thing around. He is clever, adaptable, and happy, whether hanging at home or performing as a top athlete on an agility course.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/boba.jpg",
        "key": "BknG9C2d-",
        "likes": 0,
        "name": "Boba",
      },
      Object {
        "breed": "Terrier-Schnauzer Mix",
        "color": "Salt & Pepper",
        "description": "Advancing the collected intelligence of the canine species, Cleo is smart, compassionate, and v good at snuggling. Hobbies include curling up like a donut, barking at mail-carriers, and frequent zoomies.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/cleo.jpg",
        "key": "Bkn7Hq9d-",
        "likes": 0,
        "name": "Cleo",
      },
      Object {
        "breed": "Golden Retriever",
        "color": "Yellow",
        "description": "Levi, aka 'Air Spud' is a very large adult boy who is fascinated by tennis balls & jerky treats. His personality often resembles that of a potato, though he is a very happy & big-hearted creature. Currently pursuing swimming & whatever the humans left out on the kitchen counter.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/levi.jpg",
        "key": "uTW2q9d_",
        "likes": 0,
        "name": "Levi",
      },
      Object {
        "breed": "Mutt",
        "color": "Black",
        "description": "Land seal. Tries hard af.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/harley.jpg",
        "key": "1HHChuZd8",
        "likes": 0,
        "name": "Harley",
      },
      Object {
        "breed": "Labrador Retriever",
        "color": "Chocolate",
        "description": "Loves swimming and long walks. Snores like a human.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/laughlan1.jpg",
        "key": "rWD2q1s_",
        "likes": 0,
        "name": "Laughlan",
      },
      Object {
        "breed": "Husky Dane mix",
        "color": "Black",
        "description": "Found on Lookout Mountain in Colorado, Otis is an energetic, noisy mutt that will always choose to dive into cacti in pursuit of lizards. He is currently interested in anything that moves or makes a strange sound.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/otis.jpg",
        "key": "bQo_598qR",
        "likes": 0,
        "name": "Otis",
      },
      Object {
        "breed": "Labrador Retriever",
        "color": "Black",
        "description": "A larger, fluffier, darker colored version of Laughlan. Very good at cuddling",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/boone.jpg",
        "key": "sQm_623qM",
        "likes": 0,
        "name": "Boone",
      },
      Object {
        "breed": "Lab Dane Mix",
        "color": "Yellow",
        "description": "He is big. He is loveable. He is goofy. Most of all--he is Zeus.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/zeus.jpg",
        "key": "jar_6493st",
        "likes": 0,
        "name": "Zeus",
      },
      Object {
        "breed": "Boston Terrier",
        "color": "Black & White",
        "description": "Aggresive kisser and spoiled lap dog.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/daisy.jpg",
        "key": "3TAkEh&n",
        "likes": 0,
        "name": "Daisy",
      },
      Object {
        "breed": "French Bulldog mix",
        "color": "Black & White",
        "description": "Snorts like a pig, but sophisticated like a peacock.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/penelope.jpg",
        "key": "S_eu6vPq",
        "likes": 0,
        "name": "Penelope",
      },
      Object {
        "breed": "Beagle",
        "color": "Tri-color",
        "description": "Cassie has crossed the rainbow bridge.  She was loyal, fat, and always hungry, like her owner.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/cassie.jpg",
        "key": "jar_6492ux",
        "likes": 0,
        "name": "Cassie",
      },
      Object {
        "breed": "Beagle",
        "color": "Tri-color",
        "description": "Loves to play and sleep.  Still a puppy after 14 years.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/elliemae.jpg",
        "key": "jar_8re3qux",
        "likes": 0,
        "name": "Ellie Mae",
      },
      Object {
        "breed": "Beagle",
        "color": "Tri-color",
        "description": "Mischevious food and dirty diaper bandit.  Survivor of anal-gland carcinoma (srsly).",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/annie.jpg",
        "key": "bar_9uie1278",
        "likes": 0,
        "name": "Annie",
      },
      Object {
        "breed": "Lab/Husky/Springer Spaniel/Cocker Spaniel Mix",
        "color": "Yellow",
        "description": "As lovable as he is strange! Inseparable from his brother Charlie and his people! Chases shadows and stares at corners for hours! Fiercely loyal and master snuggle haver! Allergic to everything! Confused by his own farts!",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/odie.jpg",
        "key": "ngk_4kmk6",
        "likes": 0,
        "name": "Odie",
      },
      Object {
        "breed": "Half Rat Terrier, Half Mutt",
        "color": "Yellow",
        "description": "An unstoppable bundle of joy and friendship! Loves his friends in the Seattle office. Tolerates his brother's incessant ear chewing. Wags could power a wind turbine. The happiest little skunk you'll ever meet!",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/charlie.jpg",
        "key": "l5hf_j34j",
        "likes": 0,
        "name": "Charlie",
      },
      Object {
        "breed": "Cotralian",
        "color": "Red",
        "description": "Half Mini Australian Shepard, half Cocker Spaniel. Enjoys ball, sunshine naps, and stealing his owners' socks. Living the good life in Phoenix, Arizona.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/fred.jpg",
        "key": "bnf_13kj",
        "likes": 0,
        "name": "Fred 'Baby' Diamond",
      },
      Object {
        "breed": "Braque Francais Pyrenees",
        "color": "White",
        "description": "Born January 23, 2019, our sweet and friendly 9 week old puppy will grow into a  medium-sized dog with a soft, shorthaired coat with a gentle and family friendly character. He was bred to hunt the very rugged and arid Pyrenees Mountain range on the border of France and Spain. We look forward to many years of outdoor adventures with our new best friend.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/baker.jpg",
        "key": "kr_32619",
        "likes": 0,
        "name": "Baker",
      },
      Object {
        "breed": "Yorkie",
        "color": "Tan",
        "description": "This sweet little boy is extremely loving and loves to get belly rubs. He's super quiet and ninja like unless he's meeting another dog for the first time. Loves to hoard toys and hide them.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/milo_cesar.jpg",
        "key": "lcn3_t8h1",
        "likes": 0,
        "name": "Milo",
      },
      Object {
        "breed": "Yorkie mixed",
        "color": "Black",
        "description": "He's very sweet and overprotective of his humans but friendly once he knows you. He can only eat certain foods but loves his treats, swimming, and chewing on toys. He gets around extremely well on three legs.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/madden.jpg",
        "key": "ujm_n5g8aL",
        "likes": 0,
        "name": "Madden",
      },
      Object {
        "breed": "Golden Retriever",
        "color": "Yellow",
        "description": "Winnie is a rambunctious, energetic, and loving golden retriever.  Her favorite past times are playing fetch (her absolute favorite activity), napping, or swimming in the Chesapeake Bay.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/winnie.jpg",
        "key": "ldg83_o8n4",
        "likes": 0,
        "name": "Winnie",
      },
      Object {
        "breed": "Black Mouth Cur",
        "color": "Brown and Black",
        "description": "Islay is named for a beautiful island in Scotland, her name is pronounced Eye-lah. She is a very high energy girl who loves to hunt squirrels, rabbits, and mice. On occasion she even catches one. She is obsessed with any and all food and may have already swiped your lunch off of your desk.",
        "imageUrl": "https://raw.githubusercontent.com/FormidableLabs/dogs/master/islay.jpg",
        "key": "isl15_ks84",
        "likes": 0,
        "name": "Islay",
      },
    ],
  },
}
`;

exports[`formidadog-ql server executes the likeAllDogs mutation 1`] = `
Object {
  "data": Object {
    "likeAllDogs": Array [
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 2,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
      Object {
        "likes": 1,
      },
    ],
  },
}
`;

exports[`formidadog-ql server executes the likeDog mutation 1`] = `
Object {
  "data": Object {
    "likeDog": Object {
      "likes": 1,
    },
  },
}
`;

exports[`formidadog-ql server executes the patAllDogs mutation 1`] = `
Object {
  "data": Object {
    "patAllDogs": Array [
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 2,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
      Object {
        "pats": 1,
      },
    ],
  },
}
`;

exports[`formidadog-ql server executes the patDog mutation 1`] = `
Object {
  "data": Object {
    "patDog": Object {
      "pats": 1,
    },
  },
}
`;

exports[`formidadog-ql server executes the treatAllDogs mutation 1`] = `
Object {
  "data": Object {
    "treatAllDogs": Array [
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 2,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
      Object {
        "treats": 1,
      },
    ],
  },
}
`;

exports[`formidadog-ql server executes the treatDog mutation 1`] = `
Object {
  "data": Object {
    "treatDog": Object {
      "treats": 1,
    },
  },
}
`;
